README.md
# SignalFx Tracing Library for Go

[![GitHub Release](https://img.shields.io/github/v/release/signalfx/signalfx-go-tracing)](https://github.com/signalfx/signalfx-go-tracing/releases)
[![Go Reference](https://pkg.go.dev/badge/github.com/signalfx/signalfx-go-tracing.svg)](https://pkg.go.dev/github.com/signalfx/signalfx-go-tracing)
[![go.mod](https://img.shields.io/github/go-mod/go-version/signalfx/signalfx-go-tracing)](go.mod)
[![CircleCI](https://circleci.com/gh/signalfx/signalfx-go-tracing/tree/main.svg?style=svg)](https://circleci.com/gh/signalfx/signalfx-go-tracing/tree/main)

The SignalFx Tracing Library for Go helps you automatically instrument
Go applications with instrumented library helpers and a tracer to capture
and report distributed traces to SignalFx. You can also use the library to
manually instrument Go applications with the OpenTracing API.

## Requirements and supported software

These are the requirements to use the SignalFx Tracing Library for Go:

* Go 1.12+
* [Go Modules](https://golang.org/ref/mod) as dependency management

These are the supported libraries you can instrument:

| Library | Version |
| ------- | ------- |
| [database/sql](contrib/database/sql) | Standard Library |
| [github.com/gin-gonic/gin](contrib/gin-gonic/gin) | v1.6.2 |
| [github.com/gorilla/mux](contrib/gorilla/mux) | v1.7.4 |
| [github.com/labstack/echo](contrib/labstack/echo.v4) | v4.2.1 |
| [go.mongodb.org/mongo-driver](contrib/mongodb/mongo-go-driver) | v1.3.2 |
| [net/http](contrib/net/http) | Standard Library |

Other libraries are available to instrument, but are in beta and aren't
officially supported. You can view all the libraries in the
[contrib](contrib) directory.

## Configuration values

If the default configuration values don't apply for your environment, override them before running the process you instrument.

| Code | Environment Variable | Default Value | Notes |
| ---  | ---                  | ---           | ---   |
| [WithServiceName](https://godoc.org/github.com/signalfx/signalfx-go-tracing/tracing/#WithServiceName) | `SIGNALFX_SERVICE_NAME` | `SignalFx-Tracing` | The name of the service. |
| [WithEndpointURL](https://godoc.org/github.com/signalfx/signalfx-go-tracing/tracing/#WithEndpointURL) | `SIGNALFX_ENDPOINT_URL` | `http://localhost:9080/v1/trace` | The URL to send traces to. Send spans to a Smart Agent, OpenTelemetry Collector, or a SignalFx ingest endpoint.  |
| [WithAccessToken](https://godoc.org/github.com/signalfx/signalfx-go-tracing/tracing/#WithAccessToken) | `SIGNALFX_ACCESS_TOKEN` | none | The access token for your SignalFx organization. |
| [WithGlobalTag](https://godoc.org/github.com/signalfx/signalfx-go-tracing/tracing/#WithGlobalTag) | `SIGNALFX_SPAN_TAGS` | none | Comma-separated list of tags included in every reported span. For example, "key1:val1,key2:val2". Use only string values for tags.|
| [WithRecordedValueMaxLength](https://godoc.org/github.com/signalfx/signalfx-go-tracing/tracing/#WithRecordedValueMaxLength) | `SIGNALFX_RECORDED_VALUE_MAX_LENGTH` | `1200` | The maximum number of characters for any Zipkin-encoded tagged or logged value. Behaviour disabled when set to -1. |
| - | `SIGNALFX_TRACE_RESPONSE_HEADER_ENABLED` | `true` | Adds `Server-Timing` header to HTTP responses for [net/http](contrib/net/http) and [github.com/gorilla/mux](contrib/gorilla/mux) instrumentations. |

## Instrument a Go application

Follow these steps to instrument target libraries with provided instrumentors.

For more information about how to instrument a Go application, see the
[examples](https://github.com/signalfx/tracing-examples/tree/master/signalfx-tracing/signalfx-go-tracing).

1. Import `github.com/signalfx/signalfx-go-tracing` to your `go mod`.
2. Import the instrumentor for the target library you want to instrument and
replace utilities with their traced equivalents.

   Find an instrumentor for each supported target library in the [contrib](contrib)
   directory. Each instrumentor has an `example_test.go` file that demonstrates
   how to instrument a target library. Don't import examples directly in your application.
3. Enable tracing globally with
[tracing.Start()](https://godoc.org/github.com/signalfx/signalfx-go-tracing/tracing/#Start).
This creates a tracer and registers it as the OpenTracing global tracer.

## Inject trace context in HTTP requests

Link individual log entries with trace IDs and span IDs associated with corresponding events by passing metadata in HTTP reqeusts.

To inject trace context in HTTP requests, follow these steps:

1. Replace the default HTTP multiplexer:

   ```go
   import (
     httptrace "github.com/signalfx/signalfx-go-tracing/contrib/net/http"
     "github.com/signalfx/signalfx-go-tracing/ddtrace/tracer" // global tracer
     "github.com/signalfx/signalfx-go-tracing/tracing" // helper
   )
   ```

2. Create the new HTTP multiplexer and attach it to your HTTP instance:

   ```go
   mux := httptrace.NewServeMux()
   mux.HandleFunc("/", handler) // handler to your function
   // add all other routes
   http.ListenAndServe(":8888", mux)
   ```

3. Use a helper for the tracer to get the span ID and trace ID from each request:

   ```go
   func TraceIdFromCtx(ctx context.Context) (result string) {
     if span, ok := tracer.SpanFromContext(ctx); ok {
       result = tracer.TraceIDHex(span.Context())
     }
     return
   }
   ```

   It cooperates with the global tracer instance.

### Example log injection configuration

This is an example of what injecting span IDs and trace IDs can look like in your environment:

```go
package main

import (
  httptrace "github.com/signalfx/signalfx-go-tracing/contrib/net/http"
  "github.com/signalfx/signalfx-go-tracing/ddtrace/tracer" // global tracer
  "github.com/signalfx/signalfx-go-tracing/tracing" // helper

  "context"
  "fmt"
  "log"
  "net/http"
)

func TraceIdFromCtx(ctx context.Context) (result string) {
  if span, ok := tracer.SpanFromContext(ctx); ok {
    result = tracer.TraceIDHex(span.Context())
  }
  return
}

func handler(w http.ResponseWriter, r *http.Request) {
  ctx := r.Context()
  log.Print("Recevied request for ", r.URL.Path[1:], " ", TraceIdFromCtx(ctx))
}

func main() {
  tracing.Start()
  defer tracing.Stop()
  mux.HandleFunc("/", handler)
  http.ListenAndServe(":8888", mux)
}

```

## Testing

To run integration tests locally, you should set the `INTEGRATION` environment
variable. The dependencies of the integration tests are best run via Docker.
To get an idea about the versions and the set-up take a look at our
[CI config](https://github.com/signalfx/signalfx-go-tracing/blob/main/.circleci/config.yml).

The best way to run the entire test suite is using the
[CircleCI CLI](https://circleci.com/docs/2.0/local-jobs/). Simply run
`circleci build` in the repository root. Note that you might have to increase
the resources dedicated to Docker to around 4GB.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.12.0] - 2021-09-20

### Added

- Add `http.RTWithNoDebugStack` to the `net/http.RoundTripper` instrumentation. ([#177](https://github.com/signalfx/signalfx-go-tracing/pull/177))

## [1.11.0] - 2021-09-08

### Fixed

- Fix Confluent Kafka Client instrumentation module path. ([#163](https://github.com/signalfx/signalfx-go-tracing/pull/163))
- Bump [github.com/tinylib/msgp](https://github.com/tinylib/msgp) from 1.1.5 to [1.1.6](https://github.com/tinylib/msgp/releases/tag/v1.1.6). ([#153](https://github.com/signalfx/signalfx-go-tracing/pull/153))
- Fix panic when context contains noop opentracing span. ([#173](https://github.com/signalfx/signalfx-go-tracing/pull/173))

## [1.10.0] - 2021-06-30

### Added

- Add `redigo.WithSpanOptions` to the `github.com/gomodule/redigo` instrumentation. ([#157](https://github.com/signalfx/signalfx-go-tracing/pull/157))

## [1.9.3] - 2021-06-15

### Fixed

- Add leading zeros to the B3 propagation `x-b3-traceid` and `x-b3-spanid` headers so that they are always 16 characters long. ([#149](https://github.com/signalfx/signalfx-go-tracing/pull/149))

## [1.9.2] - 2021-04-19

### Fixed

- Fix deadlock and race condition in sarama instrumentation. ([#128](https://github.com/signalfx/signalfx-go-tracing/pull/128))

## [1.9.1] - 2021-04-15

### Added

- Sarama instrumentation now adds `message_bus.destination` attribute. ([#124](https://github.com/signalfx/signalfx-go-tracing/pull/124))

## [1.9.0] - 2021-04-14

### Added

- Redigo instrumentation now adds `span.kind` and `db.type` attributes. ([#121](https://github.com/signalfx/signalfx-go-tracing/pull/121))
- Sarama instrumentation now supports a `WithPeerServiceName` option. ([#119](https://github.com/signalfx/signalfx-go-tracing/pull/119))
- Zipkin translator now correctly maps `consumer` and `producer` span kind values. ([#119](https://github.com/signalfx/signalfx-go-tracing/pull/119))

### Changed

- `ext.SpanTypeMessageConsumer` and `ext.SpanTypeMessageProducer` now evaluate to `consumer` and `producer` respectively instead of `queue`. ([#119](https://github.com/signalfx/signalfx-go-tracing/pull/119))
- Sarama instrumentation requires Kafka version 0.11.0 or newer to work correctly. ([#120](https://github.com/signalfx/signalfx-go-tracing/pull/120))

### Fixed

- Sarama instrumentation correctly identifies spans. ([#120](https://github.com/signalfx/signalfx-go-tracing/pull/120))

## [1.8.0] - 2021-04-13

### Changed

- Rename `SIGNALFX_SERVER_TIMING_CONTEXT` environmental variable to `SIGNALFX_TRACE_RESPONSE_HEADER_ENABLED` and enable it by default. ([#96](https://github.com/signalfx/signalfx-go-tracing/pull/96)) ([#117](https://github.com/signalfx/signalfx-go-tracing/pull/117))

## [1.7.0] - 2021-03-25

### Changed

- Migrate from dep to Go Modules. ([#67](https://github.com/signalfx/signalfx-go-tracing/pull/67))
- Instrument `github.com/Shopify/sarama` instead of deprecated `gopkg.in/Shopify/sarama.v1`. ([#67](https://github.com/signalfx/signalfx-go-tracing/pull/67))

## [1.6.2] - 2021-03-12

### Fixed

- Fix MongoDB instrumentation. ([#65](https://github.com/signalfx/signalfx-go-tracing/pull/65))

## [1.6.1] - 2020-12-14

### Added

- Add `db.type` tag to Redis instrumentations. ([#53](https://github.com/signalfx/signalfx-go-tracing/pull/53))

## [1.6.0] - 2020-09-29

### Changed

- go-redis instrumentation will try to substitute any values from commands with `?`. ([#50](https://github.com/signalfx/signalfx-go-tracing/pull/50))

## [1.5.0] - 2020-09-17

### Added

- Add `TraceIDHex` and `SpanIDHex` functions that return IDs from `ddtrace.SpanContext`. ([#48](https://github.com/signalfx/signalfx-go-tracing/pull/48))

## [1.4.2] - 2020-09-14

### Fixed

- Fix grpc instrumentation to correctly add span.kind tags. ([#47](https://github.com/signalfx/signalfx-go-tracing/pull/47))

## [1.4.1] - 2020-09-02

### Fixed

- Fix version number returned in code.

## [1.4.0] - 2020-08-27

Limiting the size of span tags/attributesThis can be done in the following ways:

1. Set SIGNALFX_RECORDED_VALUE_MAX_LENGTH environment variable to a number to apply the setting globally.
2. Pass tracing.WithRecordedValueMaxLength(number) option to tracing.Start() function.
3. Pass tracer.WithRecordedValueMaxLength(number) option to tracer.StartSpan() function.

### Added

- Add support for SIGNALFX_RECORDED_VALUE_MAX_LENGTH. ([#42](https://github.com/signalfx/signalfx-go-tracing/pull/42))

## [1.3.1] - 2020-08-12

### Fixed

- Ensure that redundant span.kind tag is not set for Zipkin and that its kind field is always uppercase. ([#41](https://github.com/signalfx/signalfx-go-tracing/pull/41))

## [1.3.0] - 2020-08-05

In order to support echo/v4, the instrumentation needed to import echo
from the updated v4 path and also provide an alternate import path for
itself.

When using echo/v4, users must import the instrumentation middleware
from:

```go
"github.com/signalfx/signalfx-go-tracing/contrib/labstack/echo.v4"
```

instead of:

```go
"github.com/signalfx/signalfx-go-tracing/contrib/labstack/echo"
```

### Added

- Add support for echo/v4. ([#40](https://github.com/signalfx/signalfx-go-tracing/pull/40))

## [1.2.3] - 2020-07-31

### Fixed

- Fix version number returned in code.

## [1.2.2] - 2020-07-31

### Fixed

- Redis instrumentation correctly adds span.kind=client tag to database operation spans. ([#39](https://github.com/signalfx/signalfx-go-tracing/pull/39))

## [1.2.1] - 2020-07-30

### Added


tracing/tracing.go
package tracing

import (
	"os"
	"strconv"
	"strings"

	"github.com/opentracing/opentracing-go"
	"github.com/signalfx/signalfx-go-tracing/ddtrace/opentracer"
	"github.com/signalfx/signalfx-go-tracing/ddtrace/tracer"
)

const (
	signalfxServiceName            = "SIGNALFX_SERVICE_NAME"
	signalfxEndpointURL            = "SIGNALFX_ENDPOINT_URL"
	signalfxAccessToken            = "SIGNALFX_ACCESS_TOKEN"
	signalfxSpanTags               = "SIGNALFX_SPAN_TAGS"
	signalfxRecordedValueMaxLength = "SIGNALFX_RECORDED_VALUE_MAX_LENGTH"
	signalfxDropTags               = "SIGNALFX_DROP_TAGS"
)

const defaultRecordedValueMaxLength int = 1200

var defaults = map[string]string{
	signalfxServiceName: "unnamed-go-service",
	signalfxEndpointURL: "http://localhost:9080/v1/trace",
	signalfxAccessToken: "",
}

type config struct {
	serviceName string
	accessToken string
	url         string
	// Because there can be multiple global tags added via environment variable
	// or calls to WithGlobalTag, store them in the required StartOption format to
	// call tracer.Start() in the variadic format.
	globalTags []tracer.StartOption

	// flag to disable injecting library tags
	disableLibraryTags bool

	recordedValueMaxLength *int
	tagsToDrop             []string
}

// StartOption is a function that configures an option for Start
type StartOption = func(*config)

func defaultConfig() *config {
	return &config{
		serviceName:            envOrDefault(signalfxServiceName),
		accessToken:            envOrDefault(signalfxAccessToken),
		url:                    envOrDefault(signalfxEndpointURL),
		globalTags:             envGlobalTags(),
		recordedValueMaxLength: envRecordedValueMaxLength(),
		tagsToDrop:             envDropTags(),
	}
}

func envDropTags() []string {
	val := os.Getenv(signalfxDropTags)
	tags := []string{}
	for _, tag := range strings.Split(val, ",") {
		tags = append(tags, strings.TrimSpace(tag))
	}
	return tags
}

func envRecordedValueMaxLength() *int {
	val := os.Getenv(signalfxRecordedValueMaxLength)
	num, err := strconv.Atoi(val)
	if err != nil {
		num = defaultRecordedValueMaxLength
	}
	return &num
}

// envGlobalTags extract global tags from the environment variable and parses the value in the expected format
// key1:value1,
func envGlobalTags() []tracer.StartOption {
	var globalTags []tracer.StartOption
	var val string

	if val = os.Getenv(signalfxSpanTags); val == "" {
		return globalTags
	}

	tags := strings.Split(val, ",")
	for _, tag := range tags {
		// TODO: Currently this assumes "<stringb>" where "<stringa>:<stringb>" has no ":" in the
		// string. The TODO is to fix this logic to allow for "<stringb> to have colons, ":', in it.
		pair := strings.Split(tag, ":")
		if len(pair) == 2 {
			key := strings.TrimSpace(pair[0])
			value := strings.TrimSpace(pair[1])
			// Empty keys aren't valid in Zipkin.
			// https://github.com/openzipkin/zipkin-api/blob/d3324ac79d1aa8f5c6f0ea4febb299402e50480f/zipkin-jsonv2.proto#L50-L51
			if key == "" {
				continue
			}
			globalTag := tracer.WithGlobalTag(key, value)
			globalTags = append(globalTags, globalTag)
		}
	}

	return globalTags
}

// envOrDefault gets the given environment variable if set otherwise a default value.
func envOrDefault(envVar string) string {
	if val := os.Getenv(envVar); val != "" {
		return val
	}
	return defaults[envVar]
}

// WithServiceName changes the reported service name
func WithServiceName(serviceName string) StartOption {
	return func(c *config) {
		c.serviceName = serviceName
	}
}

// WithAccessToken configures the SignalFx access token to use when reporting
func WithAccessToken(accessToken string) StartOption {
	return func(c *config) {
		c.accessToken = accessToken
	}
}

// WithEndpointURL configures the URL to send traces to
func WithEndpointURL(url string) StartOption {
	return func(c *config) {
		c.url = url
	}
}

// WithoutLibraryTags prevents the tracer from injecting
// tracing library metadata as span tags.
func WithoutLibraryTags() StartOption {
	return func(c *config) {
		c.disableLibraryTags = true
	}
}

// WithGlobalTag sets a tag with the given key/value on all spans created by the
// tracer. This option may be used multiple times.
// Note: Since the underlying transport is Zipkin, only values with strings
// are accepted.
func WithGlobalTag(k string, v string) StartOption {
	return func(c *config) {
		globalTag := tracer.WithGlobalTag(k, v)
		c.globalTags = append(c.globalTags, globalTag)
	}
}

// WithRecordedValueMaxLength specifies the maximum length a tag/log value
// can have.
// Values are completely truncated when set to 0.
// Ignored when set to -1.
func WithRecordedValueMaxLength(l int) StartOption {
	return func(c *config) {
		c.recordedValueMaxLength = &l
	}
}

// WithDropTags ...
func WithDropTags(tags ...string) StartOption {
	return func(c *config) {
		c.tagsToDrop = tags
	}
}

// Start tracing globally
func Start(opts ...StartOption) {
	c := defaultConfig()
	for _, fn := range opts {
		fn(c)
	}

	startOptions := append(c.globalTags, tracer.WithServiceName(c.serviceName))
	startOptions = append(startOptions, tracer.WithZipkin(c.serviceName, c.url, c.accessToken, c.tagsToDrop))
	if c.disableLibraryTags {
		startOptions = append(startOptions, tracer.WithoutLibraryTags())
	}
	if c.recordedValueMaxLength != nil {
		startOptions = append(startOptions, tracer.WithTracerRecordedValueMaxLength(*c.recordedValueMaxLength))
	}
	tracer.Start(
		startOptions...,
	)

	opentracing.SetGlobalTracer(opentracer.New())
}

// Stop tracing globally
func Stop() {
	tracer.Stop()
	opentracing.SetGlobalTracer(&opentracing.NoopTracer{})
}
